File: gserver.py
import sys
import numpy as np
import time
import argparse
from threading import Lock
import queue
import grpc
from concurrent import futures

import mirror_pb2
import mirror_pb2_grpc


FRAMESPERBUFFER = 2048
LOOPTIMEMS = 0.046


class AudioService(mirror_pb2_grpc.AudioServiceServicer):
    def __init__(self):
        self._input_lock = Lock()
        self._output_lock = Lock()
        self._incoming_messages = {}
        self._outgoing_messages = queue.Queue()
        self._last_time = time.perf_counter()

    def _enqueue_chunk(self, chunk):
        with self._input_lock:
            self._incoming_messages[chunk.sender] = chunk

            if time.perf_counter() - self._last_time > LOOPTIMEMS:
                summed_float = np.empty([FRAMESPERBUFFER, 1], dtype=np.int16)

                for sender, chunk in self._incoming_messages.items():
                    chunk_int = np.frombuffer(chunk.data, dtype=np.int16)
                    for i in range(FRAMESPERBUFFER):
                        summed_float[i] = summed_float[i] + np.float64(chunk_int[i])
                self._incoming_messages = {}

                summed = np.clip(summed_float, np.iinfo(np.int16).min, np.iinfo(np.int16).max).tobytes()

                with self._output_lock:
                    outgoing = mirror_pb2.AudioChunk(
                        sender=chunk.name,
                        data=summed,
                        id=chunk.id)
                    self._outgoing_messages.put(outgoing)
                self._last_time = time.perf_counter()

    def SendAudio(self, request, context):
        self._enqueue_chunk(request)
        return mirror_pb2.Empty()

    def SendAudioStream(self, request_iterator, context):
        for chunk in request_iterator:
            self._enqueue_chunk(chunk)
        return mirror_pb2.Empty()

    def GetAudioStream(self, request_iterator, context):
        while not self._outgoing_messages.empty():
            yield self._outgoing_messages.get()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('ip', default='127.0.0.1')
    parser.add_argument('port', default='9999')
    args = parser.parse_args()

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    service = AudioService()
    mirror_pb2_grpc.add_AudioServiceServicer_to_server(service, server)
    server.add_insecure_port(f'{args.ip}:{args.port}')
    server.start()

    running = True
    while running:
        try:
            time.sleep(0.1)
        except KeyboardInterrupt:
            running = False
    server.stop(0)

File: gserver.cc
#include <iostream>
#include <string>
#include <queue>
#include <mutex>
#include <algorithm>

#include <grpc++/grpc++.h>
#include "mirror.grpc.pb.h"


#define FRAMESPERBUFFER 2048
#define LOOPTIMEMS 46


class AudioServiceImpl final : public AudioService::Service
{
    public:
        void enqueueChunk(const AudioChunk &chunk)
        {
            std::lock_guard<std::mutex> _(mInputLock);
            mIncomingMessages.insert(std::pair<std::string, AudioChunk>(chunk.sender(), chunk));

            if (std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - mLastTime).count() > LOOPTIMEMS)
            {
                std::cout << "summing together " << mIncomingMessages.size() << " chunks" << std::endl;
                float summed_float[FRAMESPERBUFFER] = {0.0f};

                for (std::pair<std::string, AudioChunk> currentMessage : mIncomingMessages)
                {
                    std::int16_t * messageInt = (std::int16_t *) &currentMessage.second.data()[0u];
                    for (int i = 0; i < FRAMESPERBUFFER; i++)
                    {
                        float a = summed_float[i];
                        float b= (float) messageInt[i];
                        summed_float[i] = a + b - a * b;
                    }
                }
                mIncomingMessages.clear();

                std::int16_t summed[FRAMESPERBUFFER];

                for (int i = 0; i < FRAMESPERBUFFER; i++)
                    summed[i] = std::min(std::int16_t(summed_float[i]), std::numeric_limits<std::int16_t>::max());

                AudioChunk outgoing;
                outgoing.set_id(chunk.id());
                outgoing.set_sender(chunk.sender());

                // Send out the data as a string -- that is, 2048 int16's become 4096 chars
                outgoing.set_data(std::string((const char *) summed, 2*FRAMESPERBUFFER));

                {
                    std::lock_guard<std::mutex> _out(mOutputLock);
                    mOutgoingMessages.push(outgoing);
                }
                mLastTime = std::chrono::high_resolution_clock::now();
            }
        }

        grpc::Status SendAudioStream(grpc::ServerContext *context, grpc::ServerReader<AudioChunk> *reader, Empty *empty) override
        {
            AudioChunk chunk;
            while (reader->Read(&chunk))
                enqueueChunk(chunk);
            return grpc::Status::OK;
        }

        grpc::Status SendAudio(grpc::ServerContext *context, const AudioChunk *chunk, Empty *empty) override
        {
            enqueueChunk(*chunk);
            return grpc::Status::OK;
        }

        grpc::Status GetAudioStream(grpc::ServerContext *context, const Empty *empty, grpc::ServerWriter<AudioChunk> *audioChunkStream) override
        {
            while (!mOutgoingMessages.empty())
            {
                std::lock_guard<std::mutex> _(mOutputLock);
                audioChunkStream->Write(mOutgoingMessages.front());
                mOutgoingMessages.pop();
            }
            return grpc::Status::OK;
        }

    private:
        std::mutex mInputLock;
        std::mutex mOutputLock;
        std::map<std::string, AudioChunk> mIncomingMessages;
        std::queue<AudioChunk> mOutgoingMessages;
        std::chrono::time_point<std::chrono::high_resolution_clock> mLastTime;
};

void RunServer() {
  std::string server_address("0.0.0.0:9999");
  AudioServiceImpl service;

  grpc::ServerBuilder builder;
  // Listen on the given address without any authentication mechanism.
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  // Register "service" as the instance through which we'll communicate with
  // clients. In this case it corresponds to an *synchronous* service.
  builder.RegisterService(&service);
  // Finally assemble the server.
  std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;

  // Wait for the server to shutdown. Note that some other thread must be
  // responsible for shutting down the server for this call to ever return.
  server->Wait();
}

int main(int argc, char** argv) {
  RunServer();

  return 0;
}
